### PR TITLE
[SYCLomatic] Fix the file scope device variable migration when dim>=1

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -3766,6 +3767,23 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+  case TypeLoc::Elaborated: {
+    const TypeLoc &NamedTypeLoc =
+        TYPELOC_CAST(ElaboratedTypeLoc).getNamedTypeLoc();
+    if (NamedTypeLoc.getTypeLocClass() == TypeLoc::Typedef)
+      return setTypeInfo(NamedTypeLoc);
+    break;
+  }
+  case TypeLoc::Typedef: {
+    const TypeLoc &TypedefTpyeDeclLoc = TYPELOC_CAST(TypedefTypeLoc)
+                                            .getTypedefNameDecl()
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: check nullptr

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -3766,6 +3767,23 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+                                            ->getTypeSourceInfo()
+                                            ->getTypeLoc();
+    if (DpctGlobalInfo::isInAnalysisScope(TypedefTpyeDeclLoc.getBeginLoc()) &&
+        TypedefTpyeDeclLoc.getTypeLocClass() == TypeLoc::ConstantArray)
+      return setTypeInfo(TypedefTpyeDeclLoc);
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: call setArrayInfo directly

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -3766,6 +3767,23 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
     return setTypeInfo(TYPELOC_CAST(ReferenceTypeLoc).getPointeeLoc());
+  case TypeLoc::Elaborated: {
+    const TypeLoc &NamedTypeLoc =
+        TYPELOC_CAST(ElaboratedTypeLoc).getNamedTypeLoc();
+    if (NamedTypeLoc.getTypeLocClass() == TypeLoc::Typedef)
+      return setTypeInfo(NamedTypeLoc);
+    break;
+  }
+  case TypeLoc::Typedef: {
+    const TypeLoc &TypedefTpyeDeclLoc = TYPELOC_CAST(TypedefTypeLoc)
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
    TypeLoc TypedefTpyeDeclLoc = TYPELOC_CAST(TypedefTypeLoc)
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 4]
```diff
@@ -3766,6 +3767,23 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
   case TypeLoc::RValueReference:
     IsReference = true;
     return setTypeInfo(TYPELOC_CAST(ReferenceTypeLoc).getPointeeLoc());
+  case TypeLoc::Elaborated: {
+    const TypeLoc &NamedTypeLoc =
+        TYPELOC_CAST(ElaboratedTypeLoc).getNamedTypeLoc();
+    if (NamedTypeLoc.getTypeLocClass() == TypeLoc::Typedef)
+      return setTypeInfo(NamedTypeLoc);
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: update basename

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -9515,26 +9515,55 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
+              new ReplaceText(Range.first, Range.second,
+                              buildString("(", ICE->getType(), ")",
+                                          Decl->getName(), ".get_ptr()")));
+        }
+      }
+    }
+    // 2. Handle address-of operation for dim>=1.
+    else if (const UnaryOperator *UO =
+                 dyn_cast_or_null<UnaryOperator>(Parent)) {
+      if (!Decl->hasAttr<CUDASharedAttr>() && UO->getOpcode() == UO_AddrOf) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:9549]: construct a CTTypeInfo only

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 6]
```diff
@@ -9515,26 +9515,55 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
+        std::shared_ptr<MemVarInfo> Info = MemVarInfo::buildMemVarInfo(Decl);
+        if (!Info)
+          return;
+        if (Info->getType()->isArray()) {
+          if (Info->getType()->getDimension() >= 2) {
+            // Dim >= 2
+            auto Range = GetReplRange(UO);
+            emplaceTransformation(
+                new ReplaceText(Range.first, Range.second,
+                                buildString("(", UO->getType(), ")",
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: reinterpret_cast

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 7]
```diff
@@ -9515,26 +9515,55 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
+                                            Decl->getName(), ".get_ptr()")));
+          } else {
+            // Dim == 1
+            auto Range = GetReplRange(UO);
+            emplaceTransformation(new InsertText(
+                Range.first, buildString("(", UO->getType(), ")")));
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: reinterpret_cast

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 7 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 5 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp:9549]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: TYPELOC_CAST returns a const T &

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 2 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 8]
```diff
@@ -3754,6 +3754,7 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
     return setTypeInfo(TYPELOC_CAST(QualifiedTypeLoc).getUnqualifiedLoc(),
                        NeedSizeFold);
   case TypeLoc::ConstantArray:
+    IsArray = true;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: move to setArrayInfo

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 9]
```diff
@@ -3766,6 +3767,37 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+      break;
+    const TypeLoc &TypedefTpyeDeclLoc = TND->getTypeSourceInfo()->getTypeLoc();
+    if (DpctGlobalInfo::isInAnalysisScope(TypedefTpyeDeclLoc.getBeginLoc()) &&
+        TypedefTpyeDeclLoc.getTypeLocClass() == TypeLoc::ConstantArray) {
+      IsArray = true;
+      return setArrayInfo(
+          static_cast<const ConstantArrayTypeLoc &>(TypedefTpyeDeclLoc),
+          NeedSizeFold);
+    }
+    std::string FullTypeStr = TL.getType().getAsString(
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: We can use TL in Elaborated to get namespace

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 10]
```diff
@@ -3766,6 +3767,37 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+    break;
+  }
+  case TypeLoc::Typedef: {
+    const TypedefNameDecl *TND =
+        TYPELOC_CAST(TypedefTypeLoc).getTypedefNameDecl();
+    if (!TND)
+      break;
+    if (!TND->getTypeSourceInfo())
+      break;
+    const TypeLoc &TypedefTpyeDeclLoc = TND->getTypeSourceInfo()->getTypeLoc();
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
    const TypeLoc TypedefTpyeDeclLoc = TND->getTypeSourceInfo()->getTypeLoc();
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 11]
```diff
@@ -3766,6 +3767,37 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+    if (DpctGlobalInfo::isInAnalysisScope(TypedefTpyeDeclLoc.getBeginLoc()) &&
+        TypedefTpyeDeclLoc.getTypeLocClass() == TypeLoc::ConstantArray) {
+      IsArray = true;
+      return setArrayInfo(
+          static_cast<const ConstantArrayTypeLoc &>(TypedefTpyeDeclLoc),
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: getAs

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 10 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 11 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 8 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 9 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 12]
```diff
@@ -3766,20 +3766,54 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+  case TypeLoc::Elaborated: {
+    const TypeLoc &NamedTypeLoc =
+        TYPELOC_CAST(ElaboratedTypeLoc).getNamedTypeLoc();
+    bool HasConstantArray = false;
+    if (NamedTypeLoc.getTypeLocClass() == TypeLoc::Typedef) {
+      HasConstantArray =
+          setTypedefInfo(NamedTypeLoc.getAs<TypedefTypeLoc>(), NeedSizeFold);
+    }
+    if (HasConstantArray)
+      return;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: move to 3775

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 13]
```diff
@@ -3766,20 +3766,54 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
   case TypeLoc::RValueReference:
     IsReference = true;
     return setTypeInfo(TYPELOC_CAST(ReferenceTypeLoc).getPointeeLoc());
+  case TypeLoc::Elaborated: {
+    const TypeLoc &NamedTypeLoc =
+        TYPELOC_CAST(ElaboratedTypeLoc).getNamedTypeLoc();
+    bool HasConstantArray = false;
+    if (NamedTypeLoc.getTypeLocClass() == TypeLoc::Typedef) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: if (auto xx = getas)

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 14]
```diff
@@ -3766,20 +3766,54 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
 
+bool CtTypeInfo::setTypedefInfo(const TypedefTypeLoc &TL, bool NeedSizeFold) {
+  const TypedefNameDecl *TND = TL.getTypedefNameDecl();
+  if (!TND)
+    return false;
+  if (!TND->getTypeSourceInfo())
+    return false;
+  const TypeLoc TypedefTpyeDeclLoc = TND->getTypeSourceInfo()->getTypeLoc();
+  if (DpctGlobalInfo::isInAnalysisScope(TypedefTpyeDeclLoc.getBeginLoc()) &&
+      TypedefTpyeDeclLoc.getTypeLocClass() == TypeLoc::ConstantArray) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: if (auto xx = getas)

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 15]
```diff
@@ -3766,20 +3766,54 @@ void CtTypeInfo::setTypeInfo(const TypeLoc &TL, bool NeedSizeFold) {
...
+      HasConstantArray =
+          setTypedefInfo(NamedTypeLoc.getAs<TypedefTypeLoc>(), NeedSizeFold);
+    }
+    if (HasConstantArray)
+      return;
+    break;
+  }
+  case TypeLoc::Typedef: {
+    setTypedefInfo(TYPELOC_CAST(TypedefTypeLoc), NeedSizeFold);
+    break;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:3781]: if true
  return

File: clang/lib/DPCT/AnalysisInfo.h [hunk 16]
```diff
@@ -2295,6 +2298,7 @@ class CtTypeInfo {
   std::string BaseNameWithoutQualifiers;
   std::vector<SizeInfo> Range;
   unsigned PointerLevel;
+  bool IsArray = false;
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.h]: IsArray can be replaced with !Range.empty()

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 12 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 15 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp:3781]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 13 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 14 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.h (hunk 16 above)
Comment (zhiweij1) [reply clang/lib/DPCT/AnalysisInfo.h]: done

**Review by ziranzha — APPROVED** [review]

**Pull request merged**